### PR TITLE
docs: Describe PR Code Coverage

### PR DESCRIPTION
**Pull request by def-**
For reviewing: https://github.com/MaterializeInc/materialize/blob/3878dccf196ea103267e7b34e8864e075d768335/doc/developer/code-coverage.md

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/developer/code-coverage.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Code Coverage for Pull Requests
+
+## Motivation
+
+A full [code coverage report](http://65.109.125.29/coverage_01879007-43c4-49db-a81d-a44e93dfb55a/) allows you to check for each part of the code if there is a test that currently covers it by executing the relevant lines of code. Unfortunately the motivation to increase code coverage using full reports is low, and the effort to understand why no existing test covers the code in question is high.
```
Comment (nrainer-materialize) [inline doc/developer/code-coverage.md:5]: "there is a test that currently covers"

I would like to have mentioned here for which test types we consider the coverage.

File: doc/developer/code-coverage.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+Thus to prevent regressions and new bugs we instead want to ensure that new and changed productive code in a pull request (PR) is adequately tested. This makes sense both when code is refactored as well as when new functionality is added. For each PR you or someone from the QA team can use the [Coverage](https://buildkite.com/materialize/coverage/builds) Buildkite pipeline to trigger coverage for it.
```
Comment (nrainer-materialize) [inline doc/developer/code-coverage.md:7]: ~~If one triggers just the coverage evaluation, is this an expensive step? If not, is there a reason why we don't automatically execute this step?~~

Ok, answered below 😀

File: doc/developer/code-coverage.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+Additionally you can also download the full coverage report in HTML format from the artifacts of the "Analyze code coverage for PR" step. This should only be necessary if you are interested in seeing coverage of the entire product:
+
+![Artifacts](assets/coverage-artifacts.png)
+
+## Limitations
+
+The [Coverage](https://buildkite.com/materialize/coverage/builds) Buildkite pipeline runs a subset of tests from the [Tests](https://buildkite.com/materialize/tests) pipeline, which is already automatically run against each pull request. The excluded tests are marked with a `coverage: skip` in the [`pipeline.template.yml`](https://github.com/MaterializeInc/materialize/blob/main/ci/test/pipeline.template.yml):
+
+- Cargo test: Code should be tested in a real environment using SQL/Kafka/etc., unit test can give a false sense of seceurity
```
Comment (nrainer-materialize) [inline doc/developer/code-coverage.md:29]: I don't fully agree with "unit test can give a false sense of security" and would rephrase it to something like that we also want to have the code seen tested in interaction.

**Review by nrainer-materialize — APPROVED** [review]
👍

File: doc/developer/code-coverage.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# Code Coverage for Pull Requests
+
+## Motivation
+
+A full [code coverage report](http://65.109.125.29/coverage_01879007-43c4-49db-a81d-a44e93dfb55a/) allows you to check for each part of the code if there is a test that currently covers it by executing the relevant lines of code. Unfortunately the motivation to increase code coverage using full reports is low, and the effort to understand why no existing mzcompose-based test covers the code in question is high.
```
Comment (philip-stoev) [inline doc/developer/code-coverage.md:5]: I think you should avoid linking directly to this machine. I would have suggested that you link to a buildkite artifact, but those can expire , so I do not have a suitable link to suggest as a replacement at this time.

File: doc/developer/README.md [hunk 5]
```diff
@@ -22,11 +22,14 @@ land, and then browsed as reference material as you skill up on the codebase.
 
 ## Table of contents
 
+* [change-data-capture.md](change-data-capture.md) describes our change data
+  capture (CDC) requirements.
+
 * [ci-regexp.md](ci-regexp.md) describes how to mark a CI issue with a regular
   expression to ignore it in future CI runs.
 
-* [change-data-capture.md](change-data-capture.md) describes our change data
```
Comment (philip-stoev) [inline doc/developer/README.md:28]: This page no longer exists, is that so?

File: doc/developer/code-coverage.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+## Triggering Coverage for your PR
+
+As an example let's take the *[adapter: Update privileges when an owner is updated](https://github.com/MaterializeInc/materialize/pull/18777)* pull request. To run PR code coverage against it, go to the [Coverage Buildkite pipeline](https://buildkite.com/materialize/coverage), click on "New Build" and enter the full commit hash and branch with your Github username:
+
+![Commit: 69de84971e8c16ab459c88be73870bb676dbc478, Branch: jkosh44:update-privilege-owners](assets/coverage-new-build.png)
+
+After some time your [Coverage run](https://buildkite.com/materialize/coverage/builds/26) will be finished and contain a summary of any uncovered lines in the form of a `git diff` where each removed line is not covered:
```
Comment (philip-stoev) [inline doc/developer/code-coverage.md:15]: I was thinking about this the other day. It may be easier to explain this if we used `git diff --output-indicator-old=!` , to use "!", a neutral yet alarming symbol, so that people are not confused by the "-". This will slightly reduce the cognitive load required to process the diff. (no need to do this in this PR)

**Review by philip-stoev — APPROVED** [review]

File: doc/developer/code-coverage.md (hunk 6 above)
Comment (nrainer-materialize) [reply doc/developer/code-coverage.md:15]: Great suggestion 👍, I was also thinking about replacing the `-` with something else (but was not aware that this can already be specified in the git command).

File: doc/developer/code-coverage.md (hunk 4 above)
Comment (def-) [reply doc/developer/code-coverage.md]: I also thought about that and don't have a good place to put it yet, but wanted to have a demo report somewhere. Removed the link for now until we have a better solution.

File: doc/developer/code-coverage.md (hunk 6 above)
Comment (def-) [reply doc/developer/code-coverage.md:15]: Good idea, that confused people already, will do so in follow-up.
